### PR TITLE
Core/Porting OMP to Parallel Utilities-VII

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
This PR ports the OMP Parallel Loops to Parallel Utilities

- ParallelUtilities

**Changelog**
-Replaced OMP Parallel For loops with Private variables with ParallelUtilities(TLS)
-Replaced OMP Atomic statements with Atomic Utilities
-Replaced Simple OMP for loop with Parallel Utitlities

### CODE REVIEW

File: kratos/processes/integration_values_extrapolation_to_nodes_process.cpp [hunk 1]
```diff
@@ -378,39 +368,36 @@ void IntegrationValuesExtrapolationToNodesProcess::InitializeVariables()
...
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
-        auto it_node = it_node_begin + i;
+    block_for_each(r_nodes_array, [&](Node<3>& rNode){
 
         // We initialize the doubles values
         for ( const auto p_var : mDoubleVariable) {
-            if (mExtrapolateNonHistorical) it_node->SetValue(*p_var, 0.0);
-            else it_node->FastGetSolutionStepValue(*p_var) = 0.0;
+            if (mExtrapolateNonHistorical) rNode.SetValue(*p_var, 0.0);
```
Comment (ashishdarekar) [inline kratos/processes/integration_values_extrapolation_to_nodes_process.cpp]: @philbucher I tried using Variable Utils for all 4 loops here, but was showing some error. _"No matching Member found"_

File: kratos/processes/levelset_convection_process.h [hunk 2]
```diff
@@ -198,13 +199,12 @@ class LevelSetConvectionProcess
...
 
         // Save current level set value and current and previous step velocity values
-        #pragma omp parallel for
-        for (int i_node = 0; i_node < static_cast<int>(mpDistanceModelPart->NumberOfNodes()); ++i_node){
-            const auto it_node = mpDistanceModelPart->NodesBegin() + i_node;
-            mVelocity[i_node] = it_node->FastGetSolutionStepValue(VELOCITY);
-            mVelocityOld[i_node] = it_node->FastGetSolutionStepValue(VELOCITY,1);
-            mOldDistance[i_node] = it_node->FastGetSolutionStepValue(mrLevelSetVar,1);
-        }
+        IndexPartition<std::size_t>(mpDistanceModelPart->NumberOfNodes()).for_each([&](std::size_t Index){
```
Comment (ashishdarekar) [inline kratos/processes/levelset_convection_process.h]: @philbucher Can we use _**Auxiliar_model_part_utilities**_ here??

File: kratos/processes/levelset_convection_process.h [hunk 3]
```diff
@@ -245,13 +244,12 @@ class LevelSetConvectionProcess
...
-        #pragma omp parallel for
-        for (int i_node = 0; i_node < static_cast<int>(mpDistanceModelPart->NumberOfNodes()); ++i_node){
-            auto it_node = mpDistanceModelPart->NodesBegin() + i_node;
-            it_node->FastGetSolutionStepValue(VELOCITY) = mVelocity[i_node];
-            it_node->FastGetSolutionStepValue(VELOCITY,1) = mVelocityOld[i_node];
-            it_node->FastGetSolutionStepValue(mrLevelSetVar,1) = mOldDistance[i_node];
-        }
+        IndexPartition<std::size_t>(mpDistanceModelPart->NumberOfNodes()).for_each([&](std::size_t Index){
+            auto it_node = mpDistanceModelPart->NodesBegin() + Index;
+            it_node->FastGetSolutionStepValue(VELOCITY) = mVelocity[Index];
```
Comment (ashishdarekar) [inline kratos/processes/levelset_convection_process.h]: Same .

File: kratos/processes/integration_values_extrapolation_to_nodes_process.cpp [hunk 4]
```diff
@@ -100,83 +100,77 @@ void IntegrationValuesExtrapolationToNodesProcess::ExecuteFinalizeSolutionStep()
...
+                rElem.CalculateOnIntegrationPoints(*p_var, aux_result, r_process_info);
                 for (IndexType i_gauss_point = 0; i_gauss_point < integration_points_number; ++i_gauss_point) {
                     for (IndexType i_node = 0; i_node < number_of_nodes; ++i_node) {
                         array_1d<double, 3>& aux_value = (mExtrapolateNonHistorical) ? r_this_geometry[i_node].GetValue(*p_var) : r_this_geometry[i_node].FastGetSolutionStepValue(*p_var);
                         const array_1d<double, 3>& aux_sol = node_coefficient(i_node, i_gauss_point) * aux_result[i_gauss_point];
-                        for (IndexType i_comp = 0; i_comp < 3; ++i_comp) {
-                            #pragma omp atomic
-                            aux_value[i_comp] += aux_sol[i_comp];
-                        }
+                        AtomicAddVector(aux_value, aux_sol);
```
Comment (philbucher) [inline kratos/processes/integration_values_extrapolation_to_nodes_process.cpp]:
```suggestion
                        AtomicAdd(aux_value, aux_sol);
```

this should work with my recent modifications

File: kratos/processes/integration_values_extrapolation_to_nodes_process.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/processes/integration_values_extrapolation_to_nodes_process.cpp]: what exactly was the error? (just curious)

I think we should anyway refactor to not have the `if` in every iteration
i.e. sth like 
~~~cpp
if (mExtrapolateNonHistorical) {
    block_for_each(
    // use SetValue ...
    );
} else {    
    // use FastGetSolutionStepValue...
    block_for_each(
    );
}

~~~

File: kratos/processes/levelset_convection_process.h (hunk 2 above)
Comment (ashishdarekar) [reply kratos/processes/levelset_convection_process.h]: @philbucher here?

**Comment by ashishdarekar** [comment]
@philbucher I guess there is a recent modification in the file _kratos/processes/levelset_convection_process.h_ (After https://github.com/KratosMultiphysics/Kratos/pull/7896). In that, I guess you have already added ParallelUtilities. Hence, There is a conflict between Master and My working branch. Should I merge the master???

**Comment by philbucher** [comment]
> @philbucher I guess there is a recent modification in the file _kratos/processes/levelset_convection_process.h_ (After #7896). In that, I guess you have already added ParallelUtilities. Hence, There is a conflict between Master and My working branch. Should I merge the master???

Ok that is unfortunate but happens :/

Yes you have to merge master and resolve the conflicts
Otherwise git won't let you merge the PR

**Review by philbucher — APPROVED** [review]
:+1:

**Pull request merged**